support `assets` for `with_options`

jesse-c opened the pull request:
It was missing, and it was preventing me from using assets dynamically.

From my limited familiarity with the Prefect codebase, and my exploration here, `asset_deps` is on the base `Task` as it's used in the Prefect engine for at least visualisations.

FIXES #18352

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [x] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [x] If this pull request adds functions or classes, it includes helpful docstrings.

zzstoatzz commented:
hi @jesse-c - thanks for the PR!

this is making sense to me, I'm just taking a think on whether we can avoid `cast` like this

jesse-c commented:
@zzstoatzz: That sounds reasonable! And yes, I was desperate to avoid duplication, haha. Happy to hear that an internal clean-up can be planned for.

I've pushed up some changes.

At first, I used the snippet you provided above (thank you!) and added you as a co-author (which I'm happy to remove too, if you'd prefer). After that though, I became convinced that the `NotSet` was only appropriate for one field, so I added a fixup commit. If that's wrong, I'll drop it. Lastly, I'm sometimes a fan of reflection/metaprogramming, so I got a solution with that that was sort of DRY—and again, happy to drop that too, and keep it simple.

jesse-c commented:
No worries! Thank you for a speedy review both before, and now.

On the unit tests failure, it looks to be a transient issue:

```
Error: buildx failed with: ERROR: failed to build: failed to solve: process "/bin/sh -c apt-get update &&     apt-get install --no-install-recommends -y     chromium     && apt-get clean && rm -rf /var/lib/apt/lists/*" did not complete successfully: exit code: 100
```
